File: Javascript/src/Decode/ir.ts
import { Code, Opcode } from './opcode';

type IrCode = {
    key: Opcode;
    num: number;
};

export class Ir {
    code: InstanceType<typeof Code>;
    irCode: IrCode[] = [];
    stack: number[] = [0];
    res: string = '';
    isRunning = false;
    /** 程序运行次数 */
    step = 0;

    get irCode_len() {
        return this.irCode.length;
    }

    constructor(code: InstanceType<typeof Code>) {
        this.code = code;
        this.convert();
    }

    convert() {
        const instruction: IrCode[] = [];
        const jump_stack = [];

        for (const e of this.code.instruction) {
            const last = instruction.slice(-1)[0];
            switch (e) {
                case Opcode.SHR: {
                    if (last?.key === Opcode.SHR) {
                        last.num += 1;
                    } else {
                        instruction.push({ key: Opcode.SHR, num: 1 });
                    }
                    break;
                }
                case Opcode.SHL: {
                    if (last?.key === Opcode.SHL) {
                        last.num += 1;
                    } else {
                        instruction.push({ key: Opcode.SHL, num: 1 });
                    }
                    break;
                }
                case Opcode.ADD: {
                    if (last?.key === Opcode.ADD) {
                        last.num = (last.num + 1) % 255;
                    } else {
                        instruction.push({ key: Opcode.ADD, num: 1 });
                    }
                    break;
                }
                case Opcode.SUB: {
                    if (last?.key === Opcode.SUB) {
                        last.num = (last.num + 1) % 255;
                    } else {
                        instruction.push({ key: Opcode.SUB, num: 1 });
                    }
                    break;
                }
                case Opcode.PutChar: {
                    instruction.push({ key: Opcode.PutChar, num: 0 });
                    break;
                }
                case Opcode.GetChar: {
                    instruction.push({ key: Opcode.GetChar, num: 0 });
                    break;
                }
                case Opcode.LB: {
                    instruction.push({ key: Opcode.LB, num: 0 });
                    jump_stack.push(instruction.length - 1); // 这里的 -1 就是去掉上一行 push 的 IR::JNZ
                    break;
                }
                case Opcode.RB: {
                    const j = jump_stack.pop();
                    if (j === undefined) {
                        throw new Error('括号指令未成对 [pop from empty list]');
                    }
                    instruction.push({ key: Opcode.RB, num: j });

                    const curr_position = instruction.length - 1;
                    const lastLB = instruction[j];
                    if (lastLB.key !== Opcode.LB) {
                        throw new Error('LB不存在，括号不成对');
                    }
                    lastLB.num = curr_position;
                    break;
                }

                default:
                    break;
            }
        }

        this.irCode = instruction;
    }

    _makeTask(
        event: {
            onFinish?: (o: string) => void;
        } = {}
    ) {
        const { onFinish } = event;

        let pc = 0;
        let sp = 0;

        let done = false;

        const task = () => {
            if (pc >= this.irCode_len) {
                done = true;
                onFinish?.(this.res);
                return;
            }

            const { key, num } = this.irCode[pc];
            switch (key) {
                case Opcode.SHR: {
                    sp += num;
                    if (sp >= this.stack.length) {
                        const expand = sp - this.stack.length + 1;
                        const temp = new Array(expand).fill(0);
                        this.stack = this.stack.concat(temp);
                    }
                    break;
                }
                case Opcode.SHL: {
                    const sub = sp - num <= 0 ? sp : num;
                    sp -= sub;
                    break;
                }
                case Opcode.ADD: {
                    const curr = (this.stack[sp] + num) % 256;
                    this.stack[sp] = curr;
                    break;
                }
                case Opcode.SUB: {
                    let res = (this.stack[sp] - num) % 256;
                    if (res < 0) {
                        res += 256;
                    }
                    this.stack[sp] = res;
                    break;
                }
                case Opcode.PutChar: {
                    const c = String.fromCharCode(this.stack[sp]);
                    this.write(c);
                    break;
                }
                case Opcode.GetChar: {
                    const s = prompt('请输入一个字符') || '';
                    const c = s.charCodeAt(0);
                    this.stack[sp] = c >= 0 && c <= 255 ? c : 0;
                    break;
                }
                case Opcode.LB: {
                    if (this.stack[sp] === 0x00) {
                        pc = num;
                    }

                    break;
                }
                case Opcode.RB: {
                    if (this.stack[sp] !== 0x00) {
                        pc = num;
                    }
                    break;
                }

                default:
                    break;
            }

            pc++;
            this.step++;
        };

        return {
            task,
            getResult: () => this.res,
            isDone: () => done,
            stop: () => {
                pc = 0;
                sp = 0;
                done = true;
                this.clear();
            },
            instance: this,
        };
    }

    run(
        event: {
            onCompile?: (
                o: ReturnType<InstanceType<typeof Ir>['_makeTask']>
            ) => void;
            onFinish?: (o: string) => void;
        } = {}
    ) {
        if (this.isRunning) {
            return;
        }
        this.isRunning = true;

        const { onCompile, onFinish } = event;

        const o = this._makeTask({ onFinish: onFinish });

        function _runTask() {
            requestIdleCallback(idle => {
                while (1) {
                    if (o.isDone()) {
                        break;
                    }

                    // 10ms 留给最后一次任务的执行时间
                    if (idle.timeRemaining() > 0) {
                        // 还有空闲时间，执行任务
                        o.task();
                    } else {
                        // 否则暂停，并在下一次空闲注册任务
                        onCompile?.(o);
                        _runTask();
                        break;
                    }
                }
            });
        }

        _runTask();
        return o;
    }

    write(s: string) {
        // console.log('s', s);
        this.res += s;
    }

    clear() {
        this.stack = [0];
        this.res = '';
    }
}

File: Rust/src/ir.rs
use super::opcode;

use std::fmt;

pub enum IR {
    SHR(u32),
    SHL(u32),
    ADD(u8),
    SUB(u8),
    PutChar,
    GetChar,
    JIZ(u32), // jump if zero
    JNZ(u32), // jump if not zero
}
impl fmt::Display for IR {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            IR::SHR(n) => write!(f, "SHR {}", n),
            IR::SHL(n) => write!(f, "SHL {}", n),
            IR::ADD(v) => write!(f, "ADD {}", v),
            IR::SUB(v) => write!(f, "SUB {}", v),
            IR::PutChar => write!(f, "PutChar"),
            IR::GetChar => write!(f, "GetChar"),
            IR::JIZ(addr) => write!(f, "JIZ {}", addr),
            IR::JNZ(addr) => write!(f, "JNZ {}", addr),
        }
    }
}

pub struct Code {
    pub instruction: Vec<IR>,
}

impl Code {
    pub fn from(instr: Vec<opcode::Opcode>) -> Result<Self, Box<dyn std::error::Error>> {
        let mut instruction: Vec<IR> = Vec::new();
        let mut jump_stack: Vec<u32> = Vec::new();

        for e in instr {
            let last = instruction.last_mut();
            match e {
                // 因为 instruction 也是顺序填入的
                // 如果当前 instruction 最后一个是 SHR，那么也就是和当前相同
                // 那么就可以“折叠”成一个 SHR 指令
                opcode::Opcode::SHR => match last {
                    Some(IR::SHR(x)) => {
                        *x += 1;
                    }
                    _ => {
                        instruction.push(IR::SHR(1));
                    }
                },
                opcode::Opcode::SHL => match last {
                    Some(IR::SHL(x)) => {
                        *x += 1;
                    }
                    _ => {
                        instruction.push(IR::SHL(1));
                    }
                },
                opcode::Opcode::ADD => match last {
                    Some(IR::ADD(x)) => {
                        let (b, _) = x.overflowing_add(1);
                        *x = b;
                    }
                    _ => {
                        instruction.push(IR::ADD(1));
                    }
                },
                opcode::Opcode::SUB => match last {
                    Some(IR::SUB(x)) => {
                        let (b, _) = x.overflowing_add(1);
                        *x = b;
                    }
                    _ => {
                        instruction.push(IR::SUB(1));
                    }
                },
                opcode::Opcode::PutChar => {
                    instruction.push(IR::PutChar);
                }
                opcode::Opcode::GetChar => {
                    instruction.push(IR::GetChar);
                }
                opcode::Opcode::LB => {
                    instruction.push(IR::JIZ(0)); // 0 仅填充 无意义
                    jump_stack.push((instruction.len() - 1) as u32); // 这里的 -1 就是去掉上一行 push 的 IR::JNZ
                }
                opcode::Opcode::RB => {
                    // 括号一定是成对出现的，那么 jump_stack 中的上一个一定是 LB
                    let j = jump_stack
                        .pop()
                        .ok_or("括号指令未成对 [pop from empty list]")?;
                    instruction.push(IR::JNZ(j));

                    let curr_position = instruction.len() - 1; // 当前 IR::JNZ 的位置

                    // 这里就是上一个 LB 的位置，将无意义的 0 修改为 RB 的位置
                    match &mut instruction[j as usize] {
                        IR::JIZ(x) => {
                            *x = curr_position as u32;
                        }
                        _ => unreachable!(),
                    }
                }
            }
        }

        Ok(Code { instruction })
    }
}
